Use cacert in Nextcloud

provokateurin opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

Replaces https://github.com/NixOS/nixpkgs/pull/405873.

I tested this on my homelab test VM and the self-signed certificates are now happily accepted when set through `security.pki.certificateFiles`.
I also want to add a test for this, but don't have the time to do that right now, therefore I want to keep the PR in draft until that is done.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

provokateurin committed:
nixos/nextcloud: Disable integrity check

provokateurin committed:
nextcloud*: Replace default certificate bundle with cacert bundle

provokateurin committed:
nixos/nextcloud: Override cacert package with config.security.pki.caBundle

Ma27 commented:
>  I also want to add a test for this, but don't have the time to do that right now, therefore I want to keep the PR in draft until that is done.

Hmm, maybe it's sufficient to expose the minio (or whatever we use) in the object-store test behind TLS and use the test CA we have in nixos/tests? If this works with verification enabled and the CA added to the system-wide bundle of the Nextcloud server, that's it I think.

Ma27 commented on pkgs/servers/nextcloud/default.nix:
```diff
@@ -32,6 +33,10 @@ let
         inherit packages;
       };
 
+      patchPhase = ''
```
`postPatch`, please.

Overriding `patchPhase` means that `patches = []` is not applied anymore. We don't have any in here, but it's not too uncommon for people to do `nextcloud.overrideAttrs (_: { patches = [ custom-patches ]; })`.

Ma27 commented on nixos/modules/security/ca.nix:
```diff
@@ -111,7 +119,12 @@ in
+    {
+      security.pki = {
+        inherit cacertPackage;
+        inherit caBundle;
```
(nit) you can also do `inherit cacertPackage caBundle;`.
Normally, I wouldn't point out such things, but there's another remark, so a second iteration is necessary anyways.

Ma27 commented on nixos/modules/security/ca.nix:
```diff
@@ -94,6 +94,14 @@ in
+      readOnly = true;
+      description = ''
+        (Read-only) the path to the final package containing the bundle of certificate authorities as a single file.
+      '';
```
maybe add a `defaultText` with the expression from above so that it's clear in the manual what this is?

Ma27 reviewed (commented):
I have an almost-working test btw ;-) 

---

Paging cacert maintainers @fpletz @lukegb

provokateurin commented:
> I have an almost-working test

Feel free to push :)

I will do the other adjustments on the weekend or next week.

Ma27 committed:
nixos/nextcloud: verify that Nextcloud is using the system-wide CA

Nextcloud manages the CA bundle on its own by default, but we patch this
out and replace it with the system-wide bundle.

Since this was originally designed for the objectstore feature, this
test ensures that an S3 behind a reverse proxy with TLS and its own CA
works fine.

Ma27 commented:
Done!

> I will do the other adjustments on the weekend or next week.

No worries.
We may miss the official branch-off, but IMHO this is perfectly reasonable to backport.

provokateurin commented:
@Ma27 If you don't mind you can also do the changes from the review, I agree with all of them. Maybe we can merge it before the branch-off then :)

provokateurin commented:
@ma27 can you fix the formatting?

Ma27 commented:
@provokateurin I just realized that we can directly use the caBundle from the CA module by making this a parameter for the nextcloud package. We also need cacert, but this doesn't even get evaluated when using the module because laziness.

I like this even more because the option feels a little pointless given that the relevant path is already exposed. Wdyt?

provokateurin commented:
Ah this solution is a lot cleaner, but the commit message was not adjusted and still mentions security.pki.cacertPackage.

Ma27 commented:
Good catch, fixed!

provokateurin commented:
Cool, looks ready to merge to me!

SuperSandro2000 commented:
Someone knows how to use compressDrvWeb now?

The following fails since this PR:

```nix
services.nextcoud.package = pkgs.compressDrvWeb pkgs.nextcloud31 {
  extraFindOperands = ''-not -iregex ".*(\/apps\/.*\/l10n\/).*"'';
};
```

```
       (stack trace truncated; use '--show-trace' to show the full, detailed trace)

       error: attribute 'override' missing
       at /nix/store/aj454c0f7biyfxzpinmw28y90377x9jv-source/nixos/modules/services/web-apps/nextcloud.nix:13:21:
           12|
           13|   overridePackage = cfg.package.override {
             |                     ^
           14|     inherit (config.security.pki) caBundle;
```
